### ISSUE TITLE
java.io.EOFException when listening onCandlestickEvent for more then 60 min

### ISSUE THREAD

**Issue by chiupakabra** [binance-exchange/binance-java-api#78]
I cant seem to be able to listen for candlestick events for longer then about 1h

```
at com.binance.api.client.impl.BinanceApiWebSocketListener.onFailure(BinanceApiWebSocketListener.java:51)
at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:543)
at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:203)
at okhttp3.RealCall$AsyncCall.execute(RealCall.java:141)
at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.EOFException
at okio.RealBufferedSource.require(RealBufferedSource.java:60)
at okio.RealBufferedSource.readByte(RealBufferedSource.java:73)
at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.java:113)
at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:97)
at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:262)
at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:201)
... 5 more

```

Do I need to call keepalive on this stream?
If yes then how do I do this? I dont see any listenKey parameters here for anything other then onUserDataUpdateEvent()

```
public interface BinanceApiWebSocketClient {
  void onDepthEvent(String symbol, BinanceApiCallback<DepthEvent> callback);
  void onCandlestickEvent(String symbol, CandlestickInterval interval, BinanceApiCallback<CandlestickEvent> callback);
  void onAggTradeEvent(String symbol, BinanceApiCallback<AggTradeEvent> callback);
  void onUserDataUpdateEvent(String listenKey, BinanceApiCallback<UserDataUpdateEvent> callback);
  void onAllMarketTickersEvent(BinanceApiCallback<List<AllMarketTickersEvent>> callback);
}

```

**Comment by joaopsilva** [comment]
Yes, you need to call keepalive, which is available as part of the Rest API client. Please see an example here: https://github.com/binance-exchange/binance-java-api#start-user-data-stream-keepalive-and-close-data-stream.

**Comment by Kamitonishe** [comment]
> If yes then how do I do this? I dont see any listenKey parameters here for anything other then onUserDataUpdateEvent()
I also dont understand this. I think in candlesticks we need something like String.format("%s@kline_%s", symbol, interval.getIntervalId()); as listenKey.
But I'm not sure.

**Comment by chiupakabra** [comment]
I actually tried with the listenKey but that doesnt have any effect on events that dont accept listenKey
I have also tried with calling keepalive with String.format("%s@kline_%s", symbol, interval.getIntervalId()); as listenKey but endpoint returns error

The streams close randomly, it can be anything from 60 min to many hours if not days, and mostly small percentage of them. So I believe that listenKey has no effect whatsoever on regular events and randonly closing streams is just a normal network behaviour since the connection is open for a long periods of time and its not unheard of connections dying or being killed somewhere in the middle, you just need to handle reconnecting.

I was able to handle reconnecting with recent addition of onFalirure() to the callback, however there is an issue I came across that if I reuse the same websocket client I dont get new events which is weird. I will open a new ticket for this
